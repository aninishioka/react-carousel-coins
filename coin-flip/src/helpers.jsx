function getRandomCoin() {
  const coinNum = Math.floor(Math.random() * 2);

  return coinNum;
}

export {getRandomCoin};