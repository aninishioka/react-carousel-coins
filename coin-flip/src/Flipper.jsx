import { useState } from "react";
import Coin from "./Coin";
import { getRandomCoin } from "./helpers";

const HEADS_IMAGE_URL = "https://upload.wikimedia.org/wikipedia/commons/c/cf/State_hood_quarter_Obv_Unc.JPG"
const TAILS_IMAGE_URL = "https://upload.wikimedia.org/wikipedia/en/1/1a/Miles_%22Tails%22_Prower_Sonic_and_All-Stars_Racing_Transformed.png"

function Flipper() {
  const [coin, setCoin] = useState();
  const [flipCounts, setFlipCounts] = useState({
    total: 0,
    heads: 0,
    tails: 0
  });

  function coinFlip() {
    const coinNum = getRandomCoin();

    setCoin((coinNum === 1) ? HEADS_IMAGE_URL : TAILS_IMAGE_URL);
    setFlipCounts(fC => {
      const newCounts = {...fC};
      newCounts.total++;
      newCounts.heads = (coinNum === 1) ? newCounts.heads + 1 : newCounts.heads;
      newCounts.tails = (coinNum !== 1) ? newCounts.tails + 1 : newCounts.tails;
      return newCounts;
    });
  }

  return (
    <div>
      {coin && <Coin url={coin}/>}
      <button onClick={coinFlip}> FLIP ME</button>
      <h5>Out of {flipCounts.total} flips, there have been {flipCounts.heads} heads and {flipCounts.tails} tails. </h5>
    </div>
  );
}

export default Flipper;
export {HEADS_IMAGE_URL, TAILS_IMAGE_URL};