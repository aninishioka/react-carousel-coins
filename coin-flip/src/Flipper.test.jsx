import { it, expect, vi } from "vitest";

import { render, fireEvent } from "@testing-library/react";
import Flipper, { TAILS_IMAGE_URL } from "./Flipper";
import * as helpers from "./helpers";

const MockRandomCoin = vi.spyOn(helpers, "getRandomCoin");

it("doesn't render a coin image on load", function () {
  const { container } = render(
    <Flipper />
  );

  expect(container.querySelector("img")).not.toBeInTheDocument();
});

it("matches the snapshot", function () {
  const { container } = render(
    <Flipper />
  );

  expect(container).toMatchSnapshot();
});

it("properly updates text", function () {
  MockRandomCoin.mockReturnValueOnce(2);
  const { container } = render(<Flipper />);

  const flipButton = container.querySelector("button");
  fireEvent.click(flipButton);

  expect(
    container.querySelector("h5")
  ).toContainHTML("Out of 1 flips, there have been 0 heads and 1 tails.");
});

it("properly shows image", function () {
  MockRandomCoin.mockReturnValueOnce(2);
  const { container } = render(<Flipper />);

  const flipButton = container.querySelector("button");
  fireEvent.click(flipButton);

  expect(
    container.querySelector("img")
  ).toHaveAttribute("src", TAILS_IMAGE_URL);
});