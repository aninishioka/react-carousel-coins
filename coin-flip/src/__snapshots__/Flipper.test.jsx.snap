// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches the snapshot 1`] = `
<div>
  <div>
    <button>
       FLIP ME
    </button>
    <h5>
      Out of 
      0
       flips, there have been 
      0
       heads and 
      0
       tails. 
    </h5>
  </div>
</div>
`;
